[MLIR] Create GPU utils library & move distribution utils

kurapov-peter opened the pull request:
Continue the move of `warp_execute_on_lane_0` op to the gpu dialect (#116994). This patch creates a utils library in GPU and moves generic helper functions there.

Commit:
[MLIR] Create GPU utils library & move distribution utils

Commit:
fix formatting

Groverkss commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,57 @@
+    llvm::SmallVector<size_t> &indices);
+
+/// Helper to know if an op can be hoisted out of the region.
+bool canBeHoisted(Operation *op, function_ref<bool(Value)> definedOutside);
```
Note that this moves static functions from a transformation to public functions in the GPU namespace. I'm not a big fan of this approach; a function like "canBeHoisted" is weird without context. These functions were fine when they were constrained to a single transformation like VectorDistribution, but now that they span dialects, it should to be done better.

A solution we used downstream for something similar was to create an inherited pattern class specific to distribution patterns: https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.h#L28

It would work really well here as well. You could have:

```
template <typename T>
struct WarpDistributionPattern : OpRewritePattern<gpu::WarpExecuteOnLane0Op> {
   virtual LogicalResult matchAndRewrite(T op, PatternRewriter &rewriter) const;
}
```

then you don't need to duplicate the `getWarpResult` logic everywhere. You can expose all of these distribution specific functions as part of the pattern as well.

kurapov-peter replied:
Sure, looks good

Groverkss commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,57 @@
+/// Helper to know if an op can be hoisted out of the region.
+bool canBeHoisted(Operation *op, function_ref<bool(Value)> definedOutside);
+
+/// Return a value yielded by `warpOp` which statifies the filter lamdba
+/// condition and is not dead.
+OpOperand *getWarpResult(WarpExecuteOnLane0Op warpOp,
+                         const std::function<bool(Operation *)> &fn);
```
duplicated?

kurapov-peter replied:
yup, thx

Commit:
hide helper methods behind a pattern interface

Commit:
get rid of template parameter

kurapov-peter commented:
@Groverkss, I removed the template parameter for now as the base class passes the warp op to the pattern rewriter anyway. Maybe we shouldn't restrict it to the `OpPatternRewriter` but that seems like a trivial change. I also removed `canBeHoisted` since it's only used in a vector-specific utility (which should be generalized first - I'm thinking about providing a generic distributed type inference utility first and then restricting it to the single dimension case for vector for example).

Commit:
cleanup

Groverkss commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,62 @@
+namespace mlir {
+namespace gpu {
+struct WarpDistributionPattern : OpRewritePattern<WarpExecuteOnLane0Op> {
+  using OpRewritePattern<WarpExecuteOnLane0Op>::OpRewritePattern;
```
nit: add a newline here

Groverkss commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,62 @@
+protected:
+  /// Return a value yielded by `warpOp` which statifies the filter lamdba
+  /// condition and is not dead.
+  static OpOperand *getWarpResult(WarpExecuteOnLane0Op warpOp,
+                                  const std::function<bool(Operation *)> &fn);
+
+  /// Helper to create a new WarpExecuteOnLane0Op with different signature.
+  static WarpExecuteOnLane0Op moveRegionToNewWarpOpAndReplaceReturns(
+      RewriterBase &rewriter, WarpExecuteOnLane0Op warpOp,
+      ValueRange newYieldedValues, TypeRange newReturnTypes);
+
+  /// Helper to create a new WarpExecuteOnLane0Op region with extra outputs.
+  /// `indices` return the index of each new output.
+  static WarpExecuteOnLane0Op moveRegionToNewWarpOpAndAppendReturns(
+      RewriterBase &rewriter, WarpExecuteOnLane0Op warpOp,
+      ValueRange newYieldedValues, TypeRange newReturnTypes,
+      llvm::SmallVector<size_t> &indices);
+
+  /// Delinearize the given `laneId` into multiple dimensions, where each
+  /// dimension's size is determined by `originalShape` and `distributedShape`
+  /// together. This function expects the total numbers of threads needed for
+  /// distribution is equal to `warpSize`. Returns true and updates
+  /// `delinearizedIds` if so.
+  static bool delinearizeLaneId(OpBuilder &builder, Location loc,
```
Any reason these should be static? I don't think they should be static.

Groverkss commented on mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:
```diff
@@ -654,6 +533,38 @@ struct WarpOpTransferWrite : public OpRewritePattern<WarpExecuteOnLane0Op> {
+  /// Clone `writeOp` assumed to be nested under `warpOp` into a new warp
+  /// execute op with the proper return type. The new write op is updated to
+  /// write the result of the new warp execute op. The old `writeOp` is deleted.
+  static vector::TransferWriteOp cloneWriteOp(RewriterBase &rewriter,
```
This should not be static

Groverkss reviewed (commented):
This is looking much better now, thank you. A few more comments.

Groverkss commented:
> @Groverkss, I removed the template parameter for now as the base class passes the warp op to the pattern rewriter anyway. Maybe we shouldn't restrict it to the `OpPatternRewriter` but that seems like a trivial change. I also removed `canBeHoisted` since it's only used in a vector-specific utility (which should be generalized first - I'm thinking about providing a generic distributed type inference utility first and then restricting it to the single dimension case for vector for example).

Fair enough, we can do the templating later. Everything else also sounds good.

kuhar commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,62 @@
+#include <numeric>
+#include <utility>
+
+namespace mlir {
+namespace gpu {
```
```suggestion
namespace mlir::gpu {
```

kuhar commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,62 @@
+namespace mlir {
+namespace gpu {
+struct WarpDistributionPattern : OpRewritePattern<WarpExecuteOnLane0Op> {
+  using OpRewritePattern<WarpExecuteOnLane0Op>::OpRewritePattern;
```
```suggestion
  using OpRewritePattern::OpRewritePattern;
```

kuhar replied:
You could also add `using Base = WarpDistributionPattern;` so that the patterns in `VectorDistribute.cpp` can do `using Base::Base;`

kuhar commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,62 @@
+  static WarpExecuteOnLane0Op moveRegionToNewWarpOpAndAppendReturns(
+      RewriterBase &rewriter, WarpExecuteOnLane0Op warpOp,
+      ValueRange newYieldedValues, TypeRange newReturnTypes,
+      llvm::SmallVector<size_t> &indices);
```
```suggestion
      SmallVectorImpl<size_t> &indices);
```

kuhar commented on mlir/lib/Dialect/GPU/Utils/DistributionUtils.cpp:
```diff
@@ -0,0 +1,143 @@
+                          warpOp.getResultTypes().end());
+  auto yield = cast<gpu::YieldOp>(
+      warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+  llvm::SmallSetVector<Value, 32> yieldValues(yield.getOperands().begin(),
```
```suggestion
  SmallSetVector<Value, 32> yieldValues(yield.getOperands().begin(),
```

kurapov-peter replied:
this results in undef symbol, I left it as is

kuhar commented on mlir/lib/Dialect/GPU/Utils/DistributionUtils.cpp:
```diff
@@ -0,0 +1,143 @@
+WarpDistributionPattern::moveRegionToNewWarpOpAndAppendReturns(
+    RewriterBase &rewriter, WarpExecuteOnLane0Op warpOp,
+    ValueRange newYieldedValues, TypeRange newReturnTypes,
+    llvm::SmallVector<size_t> &indices) {
```
```suggestion
    SmallVectorImpl<size_t> &indices) {
```

kuhar commented on mlir/lib/Dialect/GPU/Utils/DistributionUtils.cpp:
```diff
@@ -0,0 +1,143 @@
+      warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+  llvm::SmallSetVector<Value, 32> yieldValues(yield.getOperands().begin(),
+                                              yield.getOperands().end());
+  for (auto newRet : llvm::zip(newYieldedValues, newReturnTypes)) {
```
```suggestion
  for (auto [value, type] : llvm::zip_equal(newYieldedValues, newReturnTypes)) {
```

kuhar commented on mlir/lib/Dialect/GPU/Utils/DistributionUtils.cpp:
```diff
@@ -0,0 +1,143 @@
+        return &yieldOperand;
+    }
+  }
+  return {};
```
```suggestion
  return nullptr;
```

kuhar commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,62 @@
+  /// Return a value yielded by `warpOp` which statifies the filter lamdba
+  /// condition and is not dead.
+  static OpOperand *getWarpResult(WarpExecuteOnLane0Op warpOp,
+                                  const std::function<bool(Operation *)> &fn);
```
```suggestion
                                  llvm::function_ref<bool(Operation *)> fn);
```

kuhar commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -0,0 +1,62 @@
+//===- VectorDistributionUtils.h - Distribution Utilities -------*- C++ -*-===//
```
The name is out of date

kuhar reviewed (commented):
Only some cosmetic issues from me, look good overall

Commit:
address review comments

Commit:
address more comments

kuhar commented on mlir/include/mlir/Dialect/GPU/Utils/DistributionUtils.h:
```diff
@@ -30,8 +27,9 @@ struct WarpDistributionPattern : OpRewritePattern<WarpExecuteOnLane0Op> {
-                           const std::function<bool(Operation *)> &fn) const;
+  OpOperand *
+  getWarpResult(WarpExecuteOnLane0Op warpOp,
+                const llvm::function_ref<bool(Operation *)> fn) const;
```
```suggestion
                llvm::function_ref<bool(Operation *)> fn) const;
```

kurapov-peter replied:
thx

kuhar commented on mlir/lib/Dialect/GPU/Utils/DistributionUtils.cpp:
```diff
@@ -83,7 +83,7 @@ WarpDistributionPattern::moveRegionToNewWarpOpAndAppendReturns(
 OpOperand *WarpDistributionPattern::getWarpResult(
     WarpExecuteOnLane0Op warpOp,
-    const std::function<bool(Operation *)> &fn) const {
+    const llvm::function_ref<bool(Operation *)> fn) const {
```
```suggestion
    llvm::function_ref<bool(Operation *)> fn) const {
```

kurapov-peter replied:
this looks like some funny github bug, the code was different at the time you left this comment :)

Commit:
remove const

kuhar commented on mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:
```diff
@@ -676,8 +587,8 @@ struct WarpOpTransferWrite : public OpRewritePattern<WarpExecuteOnLane0Op> {
-struct WarpOpElementwise : public OpRewritePattern<WarpExecuteOnLane0Op> {
-  using OpRewritePattern<WarpExecuteOnLane0Op>::OpRewritePattern;
+struct WarpOpElementwise : public WarpDistributionPattern {
+  using WarpDistributionPattern::WarpDistributionPattern;
```
You can do `using Base::Base;` now. Also elsewhere below.

Commit:
address review comments
